### PR TITLE
Introduce a `reduce` keyword argument for MSELoss

### PR DESCRIPTION
**Pull request by zou3519**
### Summary
When `reduce` is `False`, MSELoss will return a loss per batch element instead of summing or averaging the losses based on `size_average`. This was implemented by introducing `gradOutput` as an argument to the `updateGradInput` functions of the relevant loss functions. By default, `reduce = True` to maintain backwards compatibility.

This is the first step in implementing #264, which aims to add a `reduce` keyword to all loss functions.

The legacy NN `MSELoss`es currently ignore the `reduce` keyword for backwards compatibility (I'm assuming legacy means people don't usually use these).

### Test Plan
Unit tests: `test/run_test.sh`
I added relevant tests for MSELoss for when `reduce = False`. I also checked manually using `gradcheck` for the THNN case and THCUNN case (the unit tests do this as well):

```
import torch
import torch.nn as nn
from torch.autograd import Variable
from torch.autograd import gradcheck

loss = nn.MSELoss(reduce=False)
x = Variable(torch.randn(2,3,4,5).double(), requires_grad=True)
t = Variable(torch.randn(2,3,4,5).double(), requires_grad=False)
gradcheck(loss, (x,t))

x = Variable(torch.randn(2,3,4,5).double().cuda(), requires_grad=True)
t = Variable(torch.randn(2,3,4,5).double().cuda(), requires_grad=False)
gradcheck(loss, (x,t))
```

### CODE REVIEW

**Comment by soumith** [comment]
for future reference, never do `git pull origin master`, which will create merge commits such as `Merge branch 'master' into mse-reduce`. Instead do `git pull --rebase origin master`

File: torch/lib/THCUNN/generic/MSECriterion.cu [hunk 1]
```diff
@@ -7,12 +7,41 @@ void THNN_(MSECriterion_updateOutput)(
...
+      sum /= size;
+
+    THCTensor_(free)(state, input);
+    THCTensor_(free)(state, target);
+
+    THCTensor_(set1d)(state, output, 0, ScalarConvert<accreal, real>::to(sum));
+    return;
+  }
+
+  THCUNN_assertSameGPU(state, 3, input, target, output);
```
Comment (apaszke) [inline torch/lib/THCUNN/generic/MSECriterion.cu]: Move this check up. It should be performed even when `reduce=True`

File: torch/lib/THCUNN/generic/MSECriterion.cu [hunk 2]
```diff
@@ -21,54 +50,94 @@ void THNN_(MSECriterion_updateOutput)(
...
+      input_data, input_data+size, target_data, gradInput_data,
+      mse_updateGradInput_functor<real, accreal>(norm));
+
+    THCTensor_(free)(state, input);
+    THCTensor_(free)(state, target);
+    return;
+  }
+
+  THCUNN_check_shape(state, input, gradOutput);
+  THCUNN_assertSameGPU(state, 4, input, target, gradInput, gradOutput);
```
Comment (apaszke) [inline torch/lib/THCUNN/generic/MSECriterion.cu]: as above

File: torch/nn/_functions/thnn/auto_double_backwards.py [hunk 3]
```diff
@@ -224,11 +224,15 @@ def l1loss_double_backwards(ctx, ggI):
...
 def mseloss_double_backwards(ctx, ggI):
     size_average = ctx.additional_args[0]
+    reduce = ctx.additional_args[1]
     input, target, gO = ctx.saved_variables
-    div_factor = input.nelement() if size_average else 1
+    div_factor = input.nelement() if size_average and reduce else 1
 
     gI = ggI * (gO * 2. / div_factor).expand_as(input)
-    ggO = (ggI * (input - target)).sum() * (2. / div_factor)
+    if (reduce):
```
Comment (apaszke) [inline torch/nn/_functions/thnn/auto_double_backwards.py]: don't use parenthesis in `if` statements when they're unnecessary

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -13,9 +13,10 @@ def _assert_no_grad(variable):
 
 
 class _Loss(Module):
-    def __init__(self, size_average=True):
+    def __init__(self, size_average=True, reduce=True):
```
Comment (apaszke) [inline torch/nn/modules/loss.py]: This is going to make all loss functions accept the `reduce` argument, while only `MSELoss` really supports it. Either only add it to `MSELoss`, or patch all other functions to raise an error if it's set to `False`

File: test/common_nn.py [hunk 5]
```diff
@@ -284,8 +289,11 @@ def get_size_average(m):
         module_name='MSELoss',
         input=torch.randn(2, 3, 4, 5),
         target=torch.randn(2, 3, 4, 5),
-        reference_fn=lambda i, t, m: (i - t).abs().pow(2).sum() / (i.numel() if get_size_average(m) else 1),
+        reference_fn=lambda i, t, m:
+            ((i - t).abs().pow(2).sum() / (i.numel() if get_size_average(m) else 1))
+            if get_reduce(m) else (i - t).abs().pow(2),
```
Comment (apaszke) [inline test/common_nn.py]: nit: there's no point in doing `abs()` before `pow(2)`

File: test/common_nn.py [hunk 6]
```diff
@@ -742,15 +750,17 @@ def __call__(self, test_case):
...
             if isinstance(target, Variable):
                 target = target.data
             expected_out = self.reference_fn(deepcopy(self._unpack_input(input)),
                                              deepcopy(target), module)
             test_case.assertEqual(out, expected_out)
 
-        test_case.check_criterion_jacobian(module, input, self.target)
+        # check_criterion_jacobian only applies when loss is a scalar.
+        if get_reduce(module):
+            test_case.check_criterion_jacobian(module, input, self.target)
```
Comment (apaszke) [inline test/common_nn.py]: then use regular jacobian tests if `reduce` is `False`. Right now they are skipped entirely.

It might be better to reuse the regular `ModuleTest`  class for losses that don't reduce

File: test/common_nn.py [hunk 7]
```diff
@@ -769,13 +779,23 @@ def test_cuda(self, test_case):
...
+            gpu_output = test_case._forward_criterion(
+                gpu_module, gpu_input, gpu_target, reduce=reduce)
             test_case.assertEqual(cpu_output, gpu_output, 4e-4)
 
-            cpu_gradInput = test_case._backward_criterion(cpu_module, cpu_input, cpu_target)
-            gpu_gradInput = test_case._backward_criterion(gpu_module, gpu_input, gpu_target)
+            cpu_gradOutput = \
+                None if reduce else torch.randn(cpu_input.size()).double()
+            gpu_gradOutput = \
+                None if reduce else to_gpu(cpu_gradOutput, type_map=type_map)
```
Comment (apaszke) [inline test/common_nn.py]: This is confusing. Why do you give it a `None` grad if `reduce` is true?

File: test/test_legacy_nn.py [hunk 8]
```diff
@@ -632,10 +632,12 @@ def _forward(self, module, input):
     def _backward(self, module, input, output, grad_output):
         return module.backward(input, grad_output)
 
-    def _forward_criterion(self, criterion, input, target):
+    def _forward_criterion(self, criterion, input, target, reduce=True):
+        assert(reduce is True)
```
Comment (apaszke) [inline test/test_legacy_nn.py]: never put parenthesis around an `assert` condition. it's a statement so they're not needed, and it increases the risk of bugs (try `assert(False,)` in your interpreter)

File: test/test_nn.py [hunk 9]
```diff
@@ -248,21 +248,27 @@ def _backward(self, module, input, output, grad_output):
...
         input_tuple = input if isinstance(input, tuple) else (input,)
         for i in input_tuple:
             if i.grad is not None:
                 i.grad.data.zero_()
         args = input_tuple + (target,)
-        criterion(*args).backward()
+        if gradOutput is None:
+            criterion(*args).backward()
+        else:
+            criterion(*args).backward(gradOutput)
```
Comment (apaszke) [inline test/test_nn.py]: I think you can just always give `gradOutput` to `backward`. It will be ignored if it's `None`

File: test/test_nn.py [hunk 10]
```diff
@@ -4183,11 +4189,13 @@ def eval_constructor(*args, **kwargs):
 
 for test_params in criterion_tests + new_criterion_tests:
     name = test_params.pop('module_name')
-    test_params['constructor'] = getattr(nn, name)
+    ctor = getattr(nn, name)
+    test_params['constructor'] = ctor
```
Comment (apaszke) [inline test/test_nn.py]: why split this into two lines?

**Review by apaszke — COMMENTED** [review]
Looks good! I have a few minor comments

**Comment by zou3519** [comment]
Thanks for the review, @apaszke. Quick question about how PRs work (I'm pretty new to PRs on github): should my edits to address your comments be new commits on this branch, or should I squash them into the current commits via `git rebase -i` in the interest of keeping a clean commit history?

**Comment by apaszke** [comment]
It's best to put them in new commits, and we'll squash them all for you when merging.

File: test/common_nn.py (hunk 7 above)
Comment (zou3519) [reply test/common_nn.py]: The legacy cuda test for `MSECriterion` (`TestNN.test_MSECriterion_cuda`) also runs through this code. Right now I'm having the legacy test code [assume that `gradOutput` is `None`](https://github.com/pytorch/pytorch/pull/2878/files?diff=split#diff-b366aa5d4d0d6672aa91c10b561eaaf1R639) because I'm leaving the legacy code as-is and I don't think the legacy code supports having a `gradOutput`. 

I do agree that having `None` here is pretty confusing to read. I could change [the legacy test code](https://github.com/pytorch/pytorch/pull/2878/files?diff=split#diff-b366aa5d4d0d6672aa91c10b561eaaf1R639) to just ignore `gradOutput` instead of asserting that it's `None` -- what do you think about that?

File: test/test_nn.py (hunk 10 above)
Comment (zou3519) [reply test/test_nn.py]: `ctor` is reused below to make the constructors of the no_size_average and no_reduce tests

File: test/common_nn.py (hunk 6 above)
Comment (zou3519) [reply test/common_nn.py]: Thanks! I'll looking into reusing `ModuleTest`

File: test/test_nn.py (hunk 9 above)
Comment (fmassa) [reply test/test_nn.py]: Also nit: we use `camel_case` on python for variables. So maybe `grad_output`?

File: torch/lib/THCUNN/MSECriterion.cu [hunk 11]
```diff
@@ -26,6 +26,19 @@ struct mse_functor
 };
 
 template <typename Dtype, typename Acctype>
+struct mse_functor_dtype
+{
+  mse_functor_dtype() {}
+
+  __host__ __device__ Dtype operator()(const Dtype &x, const Dtype &y) const
+  {
+    Acctype z = ScalarConvert<Dtype, Acctype>::to(x)-y;
```
Comment (fmassa) [inline torch/lib/THCUNN/MSECriterion.cu]: I think it would have been better to use the proper `add`/`sub` operations on `Dtype` instead of converting them to `Acctype` and then converting back. Even though this might not make a difference (because `Acctype` for float is float, and not double), I still think it would be cleaner.
Same thing for the following line.

File: torch/lib/THCUNN/generic/MSECriterion.cu [hunk 12]
```diff
@@ -7,12 +7,41 @@ void THNN_(MSECriterion_updateOutput)(
...
+
+    ptrdiff_t size = THCTensor_(nElement)(state, input);
+
+    input = THCTensor_(newContiguous)(state, input);
+    target = THCTensor_(newContiguous)(state, target);
+
+    THCThrustAllocator thrustAlloc(state);
+    thrust::device_ptr<real> input_data(THCTensor_(data)(state, input));
+    thrust::device_ptr<real> target_data(THCTensor_(data)(state, target));
+    accreal sum = thrust::inner_product(
```
Comment (fmassa) [inline torch/lib/THCUNN/generic/MSECriterion.cu]: Wouldn't it be better to use [`THC_pointwiseApply3` instead](https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/generic/THCTensorMathPointwise.cu#L290)? This way we wouldn't need to convert the tensors to contiguous.

File: torch/lib/THCUNN/MSECriterion.cu (hunk 11 above)
Comment (fmassa) [reply torch/lib/THCUNN/MSECriterion.cu]: Actually, this might not work when `Dtype` is half. Maybe it might be better to use the THCNumerics here

File: torch/lib/THCUNN/generic/MSECriterion.cu (hunk 12 above)
Comment (fmassa) [reply torch/lib/THCUNN/generic/MSECriterion.cu]: Oh, just saw that this is for the `reduce=True` case. Disregard my comment. 

I know that this was like this before, but looking at it again I wonder in this case if using [`THC_reduceAll`](https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/generic/THCTensorMathReduce.cu#L142) wouldn't be better.

File: torch/lib/THCUNN/generic/MSECriterion.cu [hunk 13]
```diff
@@ -21,54 +50,94 @@ void THNN_(MSECriterion_updateOutput)(
   THCThrustAllocator thrustAlloc(state);
   thrust::device_ptr<real> input_data(THCTensor_(data)(state, input));
   thrust::device_ptr<real> target_data(THCTensor_(data)(state, target));
-  accreal sum = thrust::inner_product(
+  thrust::device_ptr<real> output_data(THCTensor_(data)(state, output));
+  thrust::transform(
```
Comment (fmassa) [inline torch/lib/THCUNN/generic/MSECriterion.cu]: nit: here you could use `THC_pointwiseApply3` I think.

File: torch/lib/THCUNN/MSECriterion.cu (hunk 11 above)
Comment (zou3519) [reply torch/lib/THCUNN/MSECriterion.cu]: Good call, I didn't know about THCNumerics

File: torch/lib/THCUNN/generic/MSECriterion.cu (hunk 12 above)
Comment (zou3519) [reply torch/lib/THCUNN/generic/MSECriterion.cu]: I think I'll leave this as-is because the code used to be like this?

Comment (fmassa) [reply torch/lib/THCUNN/generic/MSECriterion.cu]: Yeah, leave as-is. When I first commented on it, I haven't seen that this was what they had already.

**Comment by zou3519** [comment]
Is there a way to re-run the builds/tests without updating the pr?

**Comment by apaszke** [comment]
Yes. Our close-open trick :wink:

**Comment by soumith** [comment]
you can ignore the failure of `FAIL: test_fd_sharing (__main__.TestMultiprocessing)`, it's a flaky test.
